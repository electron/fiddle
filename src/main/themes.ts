import * as path from 'node:path';

import { IpcMainEvent, app, shell } from 'electron';
import * as fs from 'fs-extra';
import * as namor from 'namor';

import { ipcMainManager } from './ipc';
import { IpcEvents } from '../ipc-events';
import {
  DefaultThemes,
  FiddleTheme,
  LoadedFiddleTheme,
  defaultDark,
  defaultLight,
} from '../themes-defaults';

export const CONFIG_PATH = path.join(app.getPath('home'), '.electron-fiddle');
export const THEMES_PATH = path.join(CONFIG_PATH, 'themes');

/**
 * Read in a theme file.
 *
 * @export
 * @param {string} [name]
 * @returns {Promise<FiddleTheme | null>}
 */
export async function readThemeFile(
  name?: string,
): Promise<LoadedFiddleTheme | null> {
  if (!name || name === DefaultThemes.DARK) return defaultDark;
  if (name === DefaultThemes.LIGHT) return defaultLight;

  const file = name.endsWith('.json') ? name : `${name}.json`;
  const themePath = path.join(THEMES_PATH, file);

  try {
    const theme = await fs.readJSON(themePath);
    return {
      ...theme,
      name: theme.name || name.replace('.json', ''),
      file,
    };
  } catch (error) {
    console.warn(`Themes: Loading theme ${name} failed`, error);
    return null;
  }
}

/**
 * Create theme file and show in folder.
 *
 * @param {FiddleTheme} theme
 * @param {string} [name]
 * @returns {Promise<LoadedFiddleTheme>}
 */
export async function createThemeFile(
  theme: FiddleTheme | LoadedFiddleTheme,
  name?: string,
): Promise<LoadedFiddleTheme> {
  // Filter out file and css keys if they exist
  theme = Object.fromEntries(
    Object.entries(theme).filter(([key]) => !['file', 'css'].includes(key)),
  ) as FiddleTheme;

  name = name || namor.generate({ words: 2, numbers: 0 });

  const file = name.endsWith('.json') ? name : `${name}.json`;
  const themePath = path.join(THEMES_PATH, file);

  await fs.outputJSON(
    themePath,
    {
      ...theme,
      name,
    },
    { spaces: 2 },
  );

  shell.showItemInFolder(themePath);

  return {
    ...theme,
    name,
    file,
  };
}

/**
 * Reads and then returns all available themes.
 *
 * @returns {Promise<Array<LoadedFiddleTheme>>}
 */
export async function getAvailableThemes(): Promise<Array<LoadedFiddleTheme>> {
  const themes: Array<LoadedFiddleTheme> = [defaultDark, defaultLight];

  if (!fs.existsSync(THEMES_PATH)) {
    return themes;
  }

  try {
    const themeFiles = await fs.readdir(THEMES_PATH);

    for (const file of themeFiles) {
      if (!file.endsWith('.json')) {
        continue;
      }

      const theme = await readThemeFile(file);

      if (theme) {
        themes.push(theme);
      }
    }
  } catch (error) {
    console.warn(`Themes: Could not read available themes`, error);
  }

  return themes;
}

export async function openThemeFolder() {
  // Let any errors bubble through IPC
  await fs.ensureDir(THEMES_PATH);
  shell.showItemInFolder(THEMES_PATH);
}

export function setupThemes() {
  ipcMainManager.handle(
    IpcEvents.READ_THEME_FILE,
    (_: IpcMainEvent, name?: string) => readThemeFile(name),
  );
  ipcMainManager.handle(IpcEvents.GET_AVAILABLE_THEMES, (_: IpcMainEvent) =>
    getAvailableThemes(),
  );
  ipcMainManager.handle(
    IpcEvents.CREATE_THEME_FILE,
    (_: IpcMainEvent, newTheme: FiddleTheme, name?: string) =>
      createThemeFile(newTheme, name),
  );
  ipcMainManager.handle(IpcEvents.OPEN_THEME_FOLDER, (_: IpcMainEvent) =>
    openThemeFolder(),
  );
  ipcMainManager.on(IpcEvents.GET_THEME_PATH, (event) => {
    event.returnValue = THEMES_PATH;
  });
}
