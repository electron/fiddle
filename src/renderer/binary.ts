import * as fsType from 'fs-extra';
import * as path from 'path';
import * as extractZipType from 'extract-zip';

import { VersionState } from '../interfaces';
import { fancyImport } from '../utils/import';
import { normalizeVersion } from '../utils/normalize-version';
import { USER_DATA_PATH } from './constants';
import { removeTypeDefsForVersion } from './fetch-types';
import { AppState } from './state';

const pendingDownloads: Record<string, Promise<void>> = {};

async function downloadBinary(appState: AppState, version: string): Promise<void> {
  console.log(`Binary: Electron ${version} not present, downloading`);
  appState.versions[version].state = VersionState.downloading;

  const zipPath = await download(appState, version);
  const extractPath = getDownloadPath(version);
  console.log(
    `Binary: Electron ${version} downloaded, now unpacking to ${extractPath}`,
  );

  try {
    appState.versions[version].state = VersionState.unzipping;

    const fs = await fancyImport<typeof fsType>('fs-extra');
    await fs.mkdirp(getDownloadPath(version));
    await fs.emptyDir(extractPath);
    const electronFiles = await unzip(zipPath, extractPath);
    console.log(`Binary: Unzipped ${version}`, electronFiles);
    appState.versions[version].state = VersionState.ready;
  } catch (error) {
    console.warn(`Binary: Failure while unzipping ${version}`, error);
    appState.versions[version].state = VersionState.unknown;
  } finally {
    // This task is done, so remove it from the pending tasks list
    delete pendingDownloads[version];
  }
}

/**
 * General setup, called with a version. Is called during construction
 * to ensure that we always have or download at least one version.
 *
 * @param {string} iVersion
 * @returns {Promise<void>}
 */
export async function setupBinary(
  appState: AppState,
  iVersion: string,
): Promise<void> {
  const version = normalizeVersion(iVersion);

  // If we already have it, then we're done
  if (await getIsDownloaded(version)) {
    console.log(`Binary: Electron ${version} already downloaded.`);
    appState.versions[version].state = VersionState.ready;
    return;
  }

  // Return a promise that resolves when the download completes
  let pending = pendingDownloads[version];
  if (pending) {
    console.log(`Binary: Electron ${version} already downloading.`);
  } else {
    pending = pendingDownloads[version] = downloadBinary(appState, version);
  }
  return pending;
}

/**
 * Remove a version from disk. Does not update state. We'll try up to
 * four times before giving up if an error occurs.
 *
 * @param {string} iVersion
 * @returns {Promise<void>}
 */
export async function removeBinary(iVersion: string) {
  const version = normalizeVersion(iVersion);
  const fs = await fancyImport<typeof fsType>('fs-extra');
  let isDeleted = false;

  // utility to re-run removal functions upon failure
  // due to windows filesystem lockfile jank
  const rerunner = async (func: () => Promise<void>, counter = 1) => {
    try {
      await func();
    } catch (error) {
      console.warn(
        `Binary Manager: failed to run ${func.name} for ${version}, but failed`,
        error,
      );
      if (counter < 4) {
        console.log(`Binary Manager: Trying again to run ${func.name}`);
        await rerunner(func, counter + 1);
      }
    }
  };

  const binaryCleaner = async () => {
    if (await getIsDownloaded(version)) {
      // This is necessary since we're messing with .asar files inside
      // the Electron binaries. Electron, powering Fiddle, will try to
      // "correct" our calls, but we don't want that right here.
      process.noAsar = true;
      await fs.remove(getDownloadPath(version));
      process.noAsar = false;

      isDeleted = true;
    }
  };

  const typeDefsCleaner = async () => {
    await removeTypeDefsForVersion(version);
  };

  await rerunner(binaryCleaner);

  if (isDeleted) {
    await rerunner(typeDefsCleaner);
  }
}

/* Did we already download a given version?
 *
 * @param {string} version
 * @param {string} dir
 * @returns {boolean}
 */
export async function getIsDownloaded(
  version: string,
  dir?: string,
): Promise<boolean> {
  const expectedPath = getElectronBinaryPath(version, dir);
  const fs = await fancyImport<typeof fsType>('fs-extra');

  return fs.existsSync(expectedPath);
}

/**
 * Gets the expected path for the binary of a given Electron version
 *
 * @param {string} version
 * @param {string} dir
 * @returns {string}
 */
export function getElectronBinaryPath(
  version: string,
  dir: string = getDownloadPath(version),
): string {
  switch (process.platform) {
    case 'darwin':
      return path.join(dir, 'Electron.app/Contents/MacOS/Electron');
    case 'freebsd':
    case 'linux':
      return path.join(dir, 'electron');
    case 'win32':
      return path.join(dir, 'electron.exe');
    default:
      throw new Error(
        `Electron builds are not available for ${process.platform}`,
      );
  }
}

export function getDownloadingVersions(appState: AppState) {
  return Object.entries(appState.versions)
    .filter(([_, { state }]) => state === 'downloading')
    .map(([version, _]) => version);
}

/**
 * Returns an array of all versions downloaded to disk
 *
 * @returns {Promise<Array<string>>}
 */
export async function getDownloadedVersions(): Promise<Array<string>> {
  const fs = await fancyImport<typeof fsType>('fs-extra');
  const downloadPath = path.join(USER_DATA_PATH, 'electron-bin');
  console.log(`Binary: Checking for downloaded versions`);

  try {
    const directories = await fs.readdir(downloadPath);
    const knownVersions: Array<string> = [];

    for (const directory of directories) {
      if (await getIsDownloaded(directory)) {
        knownVersions.push(directory);
      }
    }

    return knownVersions;
  } catch (error) {
    console.warn(`Could not read known Electron versions`);
    return [];
  }
}

/**
 * Download an Electron version.
 *
 * @param {AppState} appState
 * @param {string} version
 * @returns {Promise<string>}
 */
async function download(appState: AppState, version: string): Promise<string> {
  const { download: electronDownload } = await import('@electron/get');
  const getProgressCallback = (progress: Progress) => {
    const roundedProgress = Math.round(progress.percent * 100) / 100;

    if (roundedProgress !== appState.versions[version].downloadProgress) {
      console.debug(
        `Binary: Version ${version} download progress: ${progress.percent}`,
      );
      appState.versions[version].downloadProgress = roundedProgress;
    }
  };

  if (!appState.versions[version]) {
    throw new Error(
      `Version ${version} does not exist in state, cannot download`,
    );
  }

  const zipFilePath = await electronDownload(version, {
    downloadOptions: {
      quiet: true,
      getProgressCallback,
    },
  });

  return zipFilePath;
}

/**
 * Gets the expected path for a given Electron version
 *
 * @param {string} version
 * @returns {string}
 */
function getDownloadPath(version: string): string {
  return path.join(USER_DATA_PATH, 'electron-bin', version);
}

/**
 * Unzips an electron package so that we can actually use it.
 *
 * @param {string} zipPath
 * @param {string} dir
 * @returns {Promise<void>}
 */
async function unzip(zipPath: string, dir: string): Promise<void> {
  const extract: typeof extractZipType = (await fancyImport<any>('extract-zip'))
    .default;

  process.noAsar = true;

  try {
    await extract(zipPath, { dir });
    console.log(`Binary: Unpacked!`);
  } finally {
    process.noAsar = false;
  }
}

interface Progress {
  percent: number;
  transferred: number;
  total: number;
}
