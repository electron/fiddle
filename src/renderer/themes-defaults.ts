import * as MonacoType from 'monaco-editor';

export interface FiddleTheme {
  name?: string;
  isDark?: boolean;
  editor: Partial<MonacoType.editor.IStandaloneThemeData>;
  common: {
    'fonts-common': string;
    'foreground-1': string;
    'foreground-2': string;
    'foreground-3': string;
    'background-4': string;
    'background-3': string;
    'background-2': string;
    'background-1': string;
    'border-color-2': string;
    'border-color-1': string;
    border: string;
    'button-text-color': string;
    'text-color-1': string;
    'text-color-2': string;
    'text-color-3': string;
    'error-color': string;
  };
}

export interface LoadedFiddleTheme extends FiddleTheme {
  name: string;
  file: string;
  css?: string;
}

/**
 * Fiddle's default dark theme.
 */
export const defaultDark: LoadedFiddleTheme = {
  name: 'Fiddle (Dark)',
  file: 'defaultDark',
  isDark: true,
  common: {
    'foreground-1': '#9feafa',
    'foreground-2': '#8ac7d6',
    'foreground-3': '#608291',
    'background-4': '#21232d',
    'background-3': '#2c2e3b',
    'background-2': '#1d2427',
    'background-1': '#2f3241',
    'border-color-2': '#1e2527',
    'border-color-1': '#5c5f71',
    border: '1px solid var(--border-color-1)',
    'button-text-color': '#000',
    'text-color-1': '#ffffff',
    'text-color-2': '#1e2527',
    'text-color-3': '#dcdcdc',
    'error-color': '#df3434',
    'fonts-common': `-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"`,
  },
  editor: {
    base: 'vs-dark',
    inherit: true,
    rules: [{ token: 'custom-date', foreground: '008800' }],
    colors: {
      'editor.background': '#2f3241',
    },
  },
};

export const defaultLight: LoadedFiddleTheme = {
  name: 'Fiddle (Light)',
  file: 'defaultLight',
  isDark: false,
  common: {
    'foreground-1': '#9feafa',
    'foreground-2': '#256e80',
    'foreground-3': '#608291',
    'background-4': '#fbfbfb',
    'background-3': '#fbfbfb',
    'background-2': '#d6dde0',
    'background-1': '#f5f5f5',
    'border-color-2': '#1e2527',
    'border-color-1': '#d8dae2',
    'button-text-color': '#fff',
    border: '1px solid var(--border-color-1)',
    'text-color-1': '#000000',
    'text-color-2': '#1e2527',
    'text-color-3': '#0e0e0e',
    'error-color': '#df3434',
    'fonts-common': `-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"`,
  },
  editor: {
    base: 'vs',
    inherit: true,
    rules: [{ token: 'custom-date', foreground: '#008800' }],
    colors: {},
  },
};

export enum DefaultThemes {
  DARK = 'defaultDark',
  LIGHT = 'defaultLight',
}
