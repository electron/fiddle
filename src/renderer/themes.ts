import * as path from 'path';

import * as fs from 'fs-extra';

import { CONFIG_PATH } from './constants';
import {
  DefaultThemes,
  FiddleTheme,
  LoadedFiddleTheme,
  defaultDark,
  defaultLight,
} from './themes-defaults';

export const THEMES_PATH = path.join(CONFIG_PATH, 'themes');

/**
 * Activate a given theme (or the default)
 *
 * @param {LoadedFiddleTheme} theme
 */
export function activateTheme(theme: LoadedFiddleTheme) {
  const { monaco } = window.ElectronFiddle;
  monaco.editor.defineTheme('main', theme.editor as any);
  monaco.editor.setTheme('main');
}

/**
 * Read in a theme file.
 *
 * @export
 * @param {string} [name]
 * @returns {Promise<FiddleTheme | null>}
 */
export async function readThemeFile(
  name?: string,
): Promise<LoadedFiddleTheme | null> {
  if (!name || name === DefaultThemes.DARK) return defaultDark as any;
  if (name === DefaultThemes.LIGHT) return defaultLight as any;

  const file = name.endsWith('.json') ? name : `${name}.json`;
  const themePath = path.join(THEMES_PATH, file);

  try {
    const theme = await fs.readJSON(themePath);
    return {
      ...theme,
      name: theme.name || name.replace('.json', ''),
      file,
    };
  } catch (error) {
    console.warn(`Themes: Loading theme ${name} failed`, error);
    return null;
  }
}

/**
 * Reads and then returns all available themes.
 *
 * @returns {Promise<Array<LoadedFiddleTheme>>}
 */
export async function getAvailableThemes(): Promise<Array<LoadedFiddleTheme>> {
  const themes: Array<LoadedFiddleTheme> = [
    defaultDark as any,
    defaultLight as any,
  ];

  if (!fs.existsSync(THEMES_PATH)) {
    return themes;
  }

  try {
    const themeFiles = await fs.readdir(THEMES_PATH);

    for (const file of themeFiles) {
      const theme = await readThemeFile(file);

      if (theme) {
        themes.push(theme);
      }
    }
  } catch (error) {
    console.warn(`Themes: Could not read available themes`, error);
  }

  return themes;
}

/**
 * Returns a Fiddle theme, either a default one or by checking
 * the disk for a JSON file.
 *
 * @export
 * @param {string} [name]
 * @returns {Promise<LoadedFiddleTheme>}
 */
export async function getTheme(
  name?: string | null,
): Promise<LoadedFiddleTheme> {
  console.log(`Themes: getTheme() loading ${name || 'default'}`);
  const theme = (await readThemeFile(name || undefined)) || defaultDark;

  return { ...theme, css: await getCssStringForTheme(theme) };
}

/**
 * Get the CSS string for a theme.
 *
 * @param {FiddleTheme} theme
 * @returns {Promise<string>}
 */
async function getCssStringForTheme(theme: FiddleTheme): Promise<string> {
  let cssContent = '';

  Object.keys(theme.common).forEach((key) => {
    cssContent += `    --${key}: ${theme.common[key]};\n`;
  });

  return `\n  html, body {\n${cssContent}  }\n`;
}
