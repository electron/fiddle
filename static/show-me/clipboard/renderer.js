window.addEventListener('DOMContentLoaded', () => {
  const copyButton = document.querySelector('#copy')
  const pasteButton = document.querySelector('#paste')
  const textarea = document.querySelector('textarea')
  copyButton.onclick = () => {
    window.clipboard.writeText('Hello from Electron!')
  }
  pasteButton.onclick = async () => {
    textarea.value = await window.clipboard.readText()
  }
})
