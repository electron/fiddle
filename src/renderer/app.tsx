import * as path from 'path';

import { ipcRenderer } from 'electron';
import { autorun, reaction, when } from 'mobx';

import { EditorValues, PACKAGE_NAME, SetFiddleOptions } from '../interfaces';
import { IpcEvents, WEBCONTENTS_READY_FOR_IPC_SIGNAL } from '../ipc-events';
import { PackageJsonOptions, getPackageJson } from '../utils/get-package';
import { USER_DATA_PATH } from './constants';
import { ElectronTypes } from './electron-types';
import { FileManager } from './file-manager';
import { ipcRendererManager } from './ipc';
import { RemoteLoader } from './remote-loader';
import { Runner } from './runner';
import { AppState } from './state';
import { TaskRunner } from './task-runner';
import { activateTheme, getTheme } from './themes';
import { defaultDark, defaultLight } from './themes-defaults';
import { getElectronVersions } from './versions';

// Importing styles files
import '../less/root.less';

/**
 * The top-level class controlling the whole app. This is *not* a React component,
 * but it does eventually render all components.
 *
 * @class App
 */
export class App {
  public state = new AppState(getElectronVersions());
  public fileManager = new FileManager(this.state);
  public remoteLoader = new RemoteLoader(this.state);
  public runner = new Runner(this.state);
  public readonly taskRunner: TaskRunner;
  public readonly electronTypes: ElectronTypes;

  constructor() {
    this.getEditorValues = this.getEditorValues.bind(this);

    this.taskRunner = new TaskRunner(this);

    this.electronTypes = new ElectronTypes(
      window.ElectronFiddle.monaco,
      path.join(USER_DATA_PATH, 'electron-typedef'),
      path.join(USER_DATA_PATH, 'nodejs-typedef'),
    );
  }

  private confirmReplaceUnsaved(): Promise<boolean> {
    return this.state.showConfirmDialog({
      label: `Opening this Fiddle will replace your unsaved changes. Do you want to proceed?`,
      ok: 'Open',
    });
  }

  private confirmExitUnsaved(): Promise<boolean> {
    return this.state.showConfirmDialog({
      label: 'The current Fiddle is unsaved. Do you want to exit anyway?',
      ok: 'Exit',
    });
  }

  public async replaceFiddle(
    editorValues: EditorValues,
    { filePath, gistId, templateName }: Partial<SetFiddleOptions>,
  ) {
    const { state } = this;
    const { editorMosaic } = state;

    if (editorMosaic.isEdited && !(await this.confirmReplaceUnsaved())) {
      return false;
    }

    this.state.editorMosaic.set(editorValues);

    this.state.gistId = gistId || '';
    this.state.localPath = filePath;
    this.state.templateName = templateName;

    // update menu when a new Fiddle is loaded
    ipcRenderer.send(IpcEvents.SET_SHOW_ME_TEMPLATE, templateName);

    return true;
  }

  /**
   * Retrieves the contents of all editor panes.
   *
   * @returns {EditorValues}
   */
  public async getEditorValues(
    options?: PackageJsonOptions,
  ): Promise<EditorValues> {
    const values = this.state.editorMosaic.values();

    if (options) {
      values[PACKAGE_NAME] = await getPackageJson(this.state, options);
    }

    return values;
  }

  /**
   * Initial setup call, loading Monaco and kicking off the React
   * render process.
   */
  public async setup(): Promise<void | Element | React.Component> {
    this.loadTheme(this.state.theme || '');

    const React = await import('react');
    const { render } = await import('react-dom');
    const { Dialogs } = await import('./components/dialogs');
    const { OutputEditorsWrapper } = await import(
      './components/output-editors-wrapper'
    );
    const { Header } = await import('./components/header');

    // The AppState constructor started loading a fiddle.
    // Wait for it here so the UI doesn't start life in `nonIdealState`.
    await when(() => this.state.editorMosaic.files.size !== 0);

    const className = `${process.platform} container`;
    const app = (
      <div className={className}>
        <Dialogs appState={this.state} />
        <Header appState={this.state} />
        <OutputEditorsWrapper appState={this.state} />
      </div>
    );

    const rendered = render(app, document.getElementById('app'));

    this.setupResizeListener();
    this.setupOfflineListener();
    this.setupThemeListeners();
    this.setupTitleListeners();
    this.setupUnloadListeners();
    this.setupTypeListeners();

    ipcRenderer.send(WEBCONTENTS_READY_FOR_IPC_SIGNAL);

    ipcRenderer.on(IpcEvents.SET_SHOW_ME_TEMPLATE, () => {
      ipcRenderer.send(IpcEvents.SET_SHOW_ME_TEMPLATE, this.state.templateName);
    });

    return rendered;
  }

  private setupTypeListeners() {
    const updateTypes = () =>
      this.electronTypes.setVersion(this.state.currentElectronVersion);
    reaction(
      () => this.state.version,
      () => updateTypes(),
    );
    updateTypes();
  }

  public async setupThemeListeners() {
    const setSystemTheme = (prefersDark: boolean) => {
      if (prefersDark) {
        this.state.setTheme(defaultDark.file);
      } else {
        this.state.setTheme(defaultLight.file);
      }
    };

    // match theme to system when box is ticked
    reaction(
      () => this.state.isUsingSystemTheme,
      () => {
        if (this.state.isUsingSystemTheme && !!window.matchMedia) {
          const { matches } = window.matchMedia('(prefers-color-scheme: dark)');
          setSystemTheme(matches);
        }
      },
    );

    // change theme when system theme changes
    if (!!window.matchMedia) {
      window
        .matchMedia('(prefers-color-scheme: dark)')
        .addEventListener('change', ({ matches }) => {
          if (this.state.isUsingSystemTheme) {
            setSystemTheme(matches);
          }
        });
    }
  }

  /**
   * Opens a fiddle from the specified location.
   *
   * @param {SetFiddleOptions} fiddle The fiddle to open
   */
  public async openFiddle(fiddle: SetFiddleOptions) {
    const { filePath, gistId } = fiddle;
    if (filePath) {
      await this.fileManager.openFiddle(filePath);
    } else if (gistId) {
      await this.remoteLoader.fetchGistAndLoad(gistId);
    }
  }

  /**
   * Loads theme CSS into the HTML document.
   *
   * @param {string} name
   * @returns {Promise<void>}
   */
  public async loadTheme(name: string): Promise<void> {
    const tag: HTMLStyleElement | null = document.querySelector(
      'style#fiddle-theme',
    );
    const theme = await getTheme(name);
    activateTheme(theme);

    if (tag && theme.css) {
      tag.innerHTML = theme.css;
    }

    if (theme.isDark || theme.name.includes('dark')) {
      document.body.classList.add('bp3-dark');
    } else {
      document.body.classList.remove('bp3-dark');
    }
  }

  public setupOfflineListener(): void {
    window.addEventListener('online', async () => {
      this.state.isOnline = true;
      this.state.setVersion(this.state.version);
    });
    window.addEventListener('offline', () => {
      this.state.isOnline = false;
    });
  }

  /**
   * We need to possibly recalculate the layout whenever the window
   * is resized. This method sets up the listener.
   */
  public setupResizeListener(): void {
    window.addEventListener('resize', this.state.editorMosaic.layout);
  }

  /**
   * Have document.title track state.title
   */
  public setupTitleListeners() {
    // the observables used for the title usually change in a batch,
    // so when setting document title, wait a tick to avoid flicker.
    let titleIdle: any;
    reaction(
      () => this.state.title,
      (title) => {
        clearTimeout(titleIdle);
        titleIdle = setTimeout(() => {
          document.title = title;
          titleIdle = null;
        });
      },
    );
  }

  public setupUnloadListeners() {
    autorun(async () => {
      const { state } = this;
      const { editorMosaic } = state;

      if (!editorMosaic.isEdited) {
        window.onbeforeunload = null;
        return;
      }

      window.onbeforeunload = async () => {
        if (await this.confirmExitUnsaved()) {
          // isQuitting checks if we're trying to quit the app
          // or just close the window
          if (state.isQuitting) {
            ipcRendererManager.send(IpcEvents.CONFIRM_QUIT);
          }
          window.onbeforeunload = null;
          window.close();
        }

        // return value doesn't matter, we just want to cancel the event
        return false;
      };
    });
  }
}
