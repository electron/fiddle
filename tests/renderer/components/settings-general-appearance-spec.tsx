import * as React from 'react';

import { IItemRendererProps } from '@blueprintjs/select';
import { shell } from 'electron';
import { shallow } from 'enzyme';
import * as fs from 'fs-extra';

import {
  AppearanceSettings,
  filterItem,
  renderItem,
} from '../../../src/renderer/components/settings-general-appearance';
import { getAvailableThemes } from '../../../src/renderer/themes';
import { FiddleTheme } from '../../../src/renderer/themes-defaults';
import { StateMock } from '../../mocks/mocks';

const mockThemes = [
  {
    name: 'defaultDark',
    file: 'defaultDark',
  },
];
const doNothingFunc = () => {
  // Do Nothing
};

jest.mock('fs-extra');

jest.mock('../../../src/renderer/themes', () => ({
  THEMES_PATH: '~/.electron-fiddle/themes',
  getAvailableThemes: jest.fn(),
  getTheme: () =>
    Promise.resolve({
      common: {},
    }),
}));

describe('AppearanceSettings component', () => {
  let store: StateMock;

  beforeEach(() => {
    ({ state: store } = (window as any).ElectronFiddle.app);

    (getAvailableThemes as jest.Mock).mockResolvedValue(mockThemes);
  });

  it('renders', () => {
    const wrapper = shallow(
      <AppearanceSettings
        appState={store as any}
        toggleHasPopoverOpen={doNothingFunc}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders the correct selected theme', async () => {
    store.theme = 'defaultDark';
    const wrapper = shallow(
      <AppearanceSettings
        appState={store as any}
        toggleHasPopoverOpen={doNothingFunc}
      />,
    );

    await process.nextTick;
    expect((wrapper.state() as any).selectedTheme.name).toBe('defaultDark');
  });

  it('handles a theme change', () => {
    const wrapper = shallow(
      <AppearanceSettings
        appState={store as any}
        toggleHasPopoverOpen={doNothingFunc}
      />,
    );
    const instance: any = wrapper.instance() as any;
    instance.handleChange({ file: 'defaultLight' } as any);

    expect(store.setTheme).toHaveBeenCalledWith('defaultLight');
  });

  it('toggles popover toggle event', () => {
    const toggleFunc = jest.fn();
    const wrapper = shallow(
      <AppearanceSettings
        appState={store as any}
        toggleHasPopoverOpen={toggleFunc}
      />,
    );

    // Find the button
    const button = wrapper.find('#open-theme-selector');

    // Simulate opening the theme selector
    button.simulate('click');
    expect(toggleFunc).toHaveBeenCalledTimes(1);

    // Simulate closing the theme selector
    button.simulate('click');
    expect(toggleFunc).toHaveBeenCalledTimes(2);
  });

  describe('openThemeFolder()', () => {
    it('attempts to open the folder', async () => {
      const wrapper = shallow(
        <AppearanceSettings
          appState={store as any}
          toggleHasPopoverOpen={doNothingFunc}
        />,
      );
      const instance: any = wrapper.instance() as any;
      await instance.openThemeFolder();

      expect(shell.showItemInFolder).toHaveBeenCalled();
    });

    it('handles an error', async () => {
      const wrapper = shallow(
        <AppearanceSettings
          appState={store as any}
          toggleHasPopoverOpen={doNothingFunc}
        />,
      );
      const instance: any = wrapper.instance() as any;
      (shell as any).showItemInFolder.mockImplementationOnce(() => {
        throw new Error('Bwap');
      });

      expect(await instance.openThemeFolder()).toBe(false);
    });
  });

  describe('createNewThemeFromCurrent()', () => {
    it('creates a new file from the current theme', async () => {
      const wrapper = shallow(
        <AppearanceSettings
          appState={store as any}
          toggleHasPopoverOpen={doNothingFunc}
        />,
      );
      const instance: any = wrapper.instance() as any;
      await instance.createNewThemeFromCurrent();

      expect(shell.showItemInFolder).toHaveBeenCalled();
      expect(fs.outputJSON).toHaveBeenCalled();

      const args = (fs.outputJSON as jest.Mock).mock.calls[0];
      expect(args[0].includes(`.electron-fiddle`)).toBe(true);
      expect(args[1].name).toBeDefined();
      expect(args[1].name === 'defaultDark').toBe(false);
      expect(args[1].common).toBeDefined();
      expect(args[1].file).toBeUndefined();
    });

    it('adds the newly created theme to the Themes dropdown', async () => {
      const arr: Array<FiddleTheme> = [];
      (getAvailableThemes as jest.Mock).mockResolvedValue(arr);
      (fs.outputJSON as jest.Mock).mockImplementation(
        (_, theme: FiddleTheme) => {
          arr.push(theme);
        },
      );
      const wrapper = shallow(
        <AppearanceSettings
          appState={store as any}
          toggleHasPopoverOpen={doNothingFunc}
        />,
      );
      expect(wrapper.state('themes')).toHaveLength(0);
      const instance: any = wrapper.instance() as any;
      await instance.createNewThemeFromCurrent();
      expect(wrapper.state('themes')).toHaveLength(1);
    });

    it('handles an error', async () => {
      const wrapper = shallow(
        <AppearanceSettings
          appState={store as any}
          toggleHasPopoverOpen={doNothingFunc}
        />,
      );
      const instance: any = wrapper.instance() as any;
      (shell as any).showItemInFolder.mockImplementationOnce(() => {
        throw new Error('Bwap');
      });

      const result = await instance.createNewThemeFromCurrent();
      expect(result).toBe(false);
    });
  });

  describe('filterItem()', () => {
    it('filters', () => {
      const foo = filterItem('foo', { name: 'foo' } as any);
      expect(foo).toBe(true);

      const bar = filterItem('foo', { name: 'bar' } as any);
      expect(bar).toBe(false);
    });
  });

  describe('renderItem()', () => {
    const mockItemProps: IItemRendererProps = {
      handleClick: () => ({}),
      index: 0,
      modifiers: {
        active: false,
        disabled: false,
        matchesPredicate: true,
      },
      query: '',
    };

    it('returns null for non-matching', () => {
      const result = renderItem({ name: 'foo' } as any, {
        ...mockItemProps,
        modifiers: {
          ...mockItemProps.modifiers,
          matchesPredicate: false,
        },
      });

      expect(result).toBe(null);
    });

    it('returns a MenuItem for matching', () => {
      const result = renderItem({ name: 'foo' } as any, {
        ...mockItemProps,
      });

      expect(result).toMatchSnapshot();
    });
  });
});
