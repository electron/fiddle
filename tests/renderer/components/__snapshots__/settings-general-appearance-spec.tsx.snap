// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppearanceSettings component renderItem() returns a MenuItem for matching 1`] = `
<Blueprint3.MenuItem
  active={false}
  disabled={false}
  icon="media"
  multiline={false}
  onClick={[Function]}
  popoverProps={Object {}}
  shouldDismissPopover={true}
  text={
    Array [
      "foo",
    ]
  }
/>
`;

exports[`AppearanceSettings component renders 1`] = `
<div
  className="settings-appearance"
>
  <h3>
    Appearance
  </h3>
  <Blueprint3.Checkbox
    checked={true}
    label="Sync theme with system setting"
    onChange={[Function]}
  />
  <Blueprint3.FormGroup
    disabled={true}
    inline={true}
    label="Choose your theme"
    labelFor="open-theme-selector"
  >
    <Blueprint3.Select
      disabled={true}
      filterable={true}
      itemPredicate={[Function]}
      itemRenderer={[Function]}
      items={Array []}
      noResults={
        <Blueprint3.MenuItem
          disabled={true}
          multiline={false}
          popoverProps={Object {}}
          shouldDismissPopover={true}
          text="No results."
        />
      }
      onItemSelect={[Function]}
      popoverProps={
        Object {
          "onClosed": [Function],
        }
      }
    >
      <Blueprint3.Button
        disabled={true}
        icon="tint"
        id="open-theme-selector"
        onClick={[Function]}
        text="Select a theme"
      />
    </Blueprint3.Select>
  </Blueprint3.FormGroup>
  <Blueprint3.Callout
    hidden={true}
  >
    <p>
      To add themes, add JSON theme files to
       
      <a
        id="open-theme-folder"
        onClick={[Function]}
      >
        <code>
          ~/.electron-fiddle/themes
        </code>
      </a>
      . The easiest way to get started is to clone one of the two existing themes and to add your own colors.
    </p>
    <p>
      Additionally, if you wish to import a Monaco Editor theme, pick your JSON file and Fiddle will attempt to import it.
    </p>
    <Blueprint3.Button
      icon="duplicate"
      onClick={[Function]}
      text="Create theme from current selection"
    />
    <Blueprint3.Button
      icon="document-open"
      onClick={[Function]}
      text="Add a Monaco Editor theme"
    />
  </Blueprint3.Callout>
</div>
`;
